rules.go

```diff
@@ -0,0 +1,482 @@
+		blacklistExcludeOptionsRegex: map[string]*regexp.Regexp{},
+	}
+
+	// Start parsing
+	for _, ruleStr := range rulesStr {
+		rule, err := parseRule(ruleStr)
+		switch {
+		case err == nil:
+			// Blacklist without options nor domain filter
+			if !rule.isException && len(rule.domains) == 0 && len(rule.options) == 0 {
+				ruleSet.blacklist = append(ruleSet.blacklist, rule)
+				continue
+			}
+			// Whitelist without options nor domain filter
+			if rule.isException && len(rule.domains) == 0 && len(rule.options) == 0 {
+				ruleSet.whitelist = append(ruleSet.whitelist, rule)
+				continue
+			}
+			// Blacklist without options with domain filter
+			if !rule.isException && len(rule.domains) > 0 && len(rule.options) == 0 {
+				ruleSet.blacklistDomainsNoOptions = append(ruleSet.blacklistDomainsNoOptions, rule)
+				continue
+			}
+			// Whitelist without options with domain filter
+			if rule.isException && len(rule.domains) > 0 && len(rule.options) == 0 {
+				ruleSet.whitelistDomainsNoOptions = append(ruleSet.whitelistDomainsNoOptions, rule)
+				continue
+			}
+			// Blacklist with options with domain filter
+			if !rule.isException && len(rule.domains) > 0 && len(rule.options) >= 0 {
+				ruleSet.blacklistDomainsWithOptions = append(ruleSet.blacklistDomainsWithOptions, rule)
+				continue
+			}
+			// Whitelist with options with domain filter
+			if rule.isException && len(rule.domains) > 0 && len(rule.options) >= 0 {
+				ruleSet.whitelistDomainsWithOptions = append(ruleSet.whitelistDomainsWithOptions, rule)
+				continue
+			}
+			// Blacklist with options without domain filter
+			if !rule.isException && len(rule.domains) == 0 && len(rule.options) > 0 {
+				for option, allowed := range rule.options {
+					if allowed {
+						ruleSet.blacklistIncludeOptions[option] = append(ruleSet.blacklistIncludeOptions[option], rule)
+					} else {
+						ruleSet.blacklistExcludeOptions[option] = append(ruleSet.blacklistExcludeOptions[option], rule)
+					}
+				}
+				continue
+			}
+			// Whitelist with options without domain filter
+			if rule.isException && len(rule.domains) == 0 && len(rule.options) > 0 {
+				for option, allowed := range rule.options {
+					if allowed {
+						ruleSet.whitelistIncludeOptions[option] = append(ruleSet.whitelistIncludeOptions[option], rule)
+					} else {
+						ruleSet.whitelistExcludeOptions[option] = append(ruleSet.whitelistExcludeOptions[option], rule)
+					}
+				}
+				continue
+			}
+		case errors.Is(err, ErrSkipComment),
+			errors.Is(err, ErrSkipHTML),
+			errors.Is(err, ErrUnsupportedRule):
+			logger.Info(err, ": ", strings.TrimSpace(ruleStr))
+		case errors.Is(err, ErrEmptyLine):
+			logger.Info(err)
+		default:
+			logger.Info("cannot parse rule: ", err)
+			return nil, fmt.Errorf("cannot parse rule: %w", err)
+		}
+	}
+
+	compileAllRegex(ruleSet)
+	return ruleSet, nil
+}
+
+func CombinedRegex(rules []*ruleAdBlock) *regexp.Regexp {
+	regexes := []string{}
```

Comment:
1. This function should not exist because of the comment above
2. Even if this function exist, it should not be a part of public interface
3. Even if 1-2 are not valid, please consider using https://golang.org/pkg/strings/#Builder there